cdc/generation.cc

```diff
@@ -342,18 +310,6 @@ future<cdc::generation_id_v2> make_new_cdc_generation(
     using namespace std::chrono;
     auto gen = topology_description_generator(cfg, bootstrap_tokens, tmptr, g).generate();
 
-    // If the cluster is large we may end up with a generation that contains
-    // large number of streams. This is problematic because we store the
-    // generation in a single row. For a generation with large number of rows
-    // this will lead to a row that can be as big as 32MB. This is much more
-    // than the limit imposed by commitlog_segment_size_in_mb. If the size of
-    // the row that describes a new generation grows above
-    // commitlog_segment_size_in_mb, the write will fail and the new node won't
-    // be able to join. To avoid such problem we make sure that such row is
-    // always smaller than 4MB. We do that by removing some CDC streams from
-    // each vnode if the total number of streams is too large.
-    gen = limit_number_of_streams_if_needed(std::move(gen));
-
     auto uuid = utils::make_random_uuid();
```

Comment:
This should remain, but only be triggered if the cluster is not using v2 format.

Reply:
If the cluster is not using v2 format this code will never be executed.

Reply:
Why not? isn't it used in v1 format?

Please explain.

Reply:
```
//  * Precondition: the CDC_GENERATIONS_V2 feature is enabled in the cluster.
```
(in cdc/generation.hh where the function is declared)
I should probably add a check for this precondition inside the function as well for sanity.
But you can check in the code that after this PR nobody calls `make_new_cdc_generation` without checking if the feature is enabled first.

Reply:
There are two uses of this function:
1. `bootstrap()`: new nodes will refuse to bootstrap into a cluster without this feature
2. `check_and_repair_cdc_streams`
3. `join_token_ring` restarting node but no CDC generation exists case: there's a BIG comment there and an error message if we see that that the feature is not enabled and the generation is missing.

Reply:
There are three uses of this function:
1. `bootstrap()`: new nodes will refuse to bootstrap into a cluster without this feature
2. `check_and_repair_cdc_streams`
3. `join_token_ring` restarting node but no CDC generation exists case: there's a BIG comment there and an error message if we see that that the feature is not enabled and the generation is missing.

Reply:
Not sure I understand 3, but I'd like to remove 1 if possible.

Reply:
Is it not enough that users can still bootstrap old nodes into a partially upgraded cluster?
I think new nodes refusing to bootstrap is a reasonable restriction.
- We don't recommend scaling in the middle of upgrade.
- But if you REALLY must bootstrap, bootstrap an old node.
- The restriction doesn't affect decommission/removenode.
- If you want to bootstrap a new node, finish upgrade first.

We could remove the `bootstrap()` restriction by readding back a lot of code: the old CDC generation insertion code, the old tables definitions (meaning that fresh clusters would also have these tables, although empty) etc. so new nodes bootstrapping to partially upgraded clusters would add a generation in the old format if they see the feature disabled.

But I think we should prevent users from doing "stupid" things more - such as scaling in the middle of upgrade - rather than allowing them more "stupid" things. I don't think QA ever tests those scenarios (quick glance at SCT code didn't reveal it), so why not exclude them in the first place? I'm sure we would hit unexpected bugs right now.

IF we want to start supporting such scenarios, THEN I could readd the old code/table definitions, change the restriction to an if-else (if feature enabled do that, else do that), AND of course QA would have to test those scenarios in their release cycle. Even then I would leave this restriction in our upcoming "safe mode". But why not be safe by default?

Reply:
The problem is that no one reads the recommendations, least of all the robots in automated services.

Reply:
> The problem is that no one reads the recommendations, least of all the robots in automated services.

Programmers of these robots should read them.
If they don't, they have a bug in their code.
If their automated service stumbles on this restriction, this would reveal their bug - which is good! - and force them to fix it.

From that POV having the restriction is better because it forces users to follow recommendations.

Reply:
Also, you mentioned Kubernetes autoscaler.
Kubernetes doesn't automatically understand how to manage a database. You must write a dedicated operator.
When writing this operator you obviously should follow the rules of the database.

Reply:
Consider this hypothetical scenario:
- I readd all the old code and remove the restrictions, allowing new nodes to be booted into mixed clusters.
- But I never actually check if this scenario works.

That obviously wouldn't be good and I should not do it, but consider: isn't this what we do for everything else? These scenarios are _not_ being tested - at least not regularly before each release (correct me if I'm wrong). But because I thought that we don't support and test them, I added a restriction that _rules out_ the untested scenario, allowing me to simplify the code greatly (mostly by removing a lot of legacy code). And simpler and shorter code means fewer bugs.

Handling the scenario is _a lot_ of additional code which would be executed only if someone actually uses Scylla _incorrectly_, and I doubt it would even be executed ever because it's a very specific edge case (who in their right mind tries to scale in the middle of upgrade). I see negative value in that. From the point of view of tradeoffs, isn't it much better to prevent such incorrect usage instead of keeping a lot of code, increasing the chance of bugs and reducing maintainability? Now if the user _really really_ wants to bootstrap a node into a mixed cluster (which they still shouldn't do), they _do_ have a way (boot old node), so it's not like we completely prevent them from doing the wrong thing, but at least we force them to think about it a bit (if they first try to boot a new node and see that it refuses) before they decide to do it.

cc @slivne - could you confirm what the policy about scaling clusters in the middle of upgrade is? Did such situations happen in practice?

Reply:
@kbr- you're obviously right that they should. That doesn't mean that they will. We need to make this thing less brittle.

Reply:
But doesn't an explicit restriction that tells the user "don't do that, do this instead" actually _increase_ robustness (so it makes things less brittle)? And allowing scenarios which are not regularly tested on each release decreases it?

Reply:
If you try to do something, and it tell you not to do it, then you are more robust, but also more brittle. The user was not able to accomplish what they wanted to do.

Reply:
Of course, we have to start testing mixed-cluster operations, or we end up with reduced robustness.

Reply:
OK. IF we really eventually start testing those scenarios regularly (tell me sweet lies) then I'll be happy with the change.